runtime error: slice bounds out of range [:2] with capacity 1

## Bug Report

Please answer these questions before submitting your issue. Thanks!

### 1. Minimal reproduce step (Required)
``` sql
Some unknown actions...
SELECT (SELECT tt.a FROM t1  tt LIMIT 1) aa, COUNT(DISTINCT b) FROM t1  GROUP BY aa;
```

### 2. What did you expect to see? (Required)
Work fine.

### 3. What did you see instead (Required)
```
338315:[2021/01/28 12:53:27.740 +00:00] [ERROR] [conn.go:733] ["connection running loop panic"] [conn=102981] [lastSQL="SELECT (SELECT tt.a FROM t1  tt LIMIT 1) aa, COUNT(DISTINCT b) FROM t1  GROUP BY aa"] [err="runtime error: slice bounds out of range [:2] with capacity 1"] [stack="goroutine 233298 [running]:
github.com/pingcap/tidb/server.(*clientConn).Run.func1(0x3e3b520, 0xc0030b4180, 0xc0002b7d00)
	/home/jenkins/agent/workspace/build-tidb/tidb/server/conn.go:731 +0xee
panic(0x360e0c0, 0xc002316c20)
	/usr/local/go/src/runtime/panic.go:679 +0x1b2
github.com/pingcap/tidb/planner/core.(*PlanBuilder).rewriteExprNode.func1(0xc0030ccd00, 0x2)
	/home/jenkins/agent/workspace/build-tidb/tidb/planner/core/expression_rewriter.go:179 +0x1f7
github.com/pingcap/tidb/planner/core.(*PlanBuilder).rewriteExprNode(0xc00246d4a0, 0xc0030ccd00, 0x3e70f20, 0xc0027425a0, 0xc0030ccc01, 0x3ebbbc0, 0xc0029be780, 0x3e8bb00, 0xc0030ccc30, 0x0, ...)
	/home/jenkins/agent/workspace/build-tidb/tidb/planner/core/expression_rewriter.go:209 +0x190
github.com/pingcap/tidb/planner/core.(*PlanBuilder).rewriteWithPreprocess(0xc00246d4a0, 0x3e3b4a0, 0xc000118008, 0x3e70f20, 0xc0027425a0, 0x3e8bb00, 0xc0030ccc30, 0x0, 0x0, 0x1, ...)
	/home/jenkins/agent/workspace/build-tidb/tidb/planner/core/expression_rewriter.go:142 +0x163
github.com/pingcap/tidb/planner/core.(*PlanBuilder).rewrite(0xc00246d4a0, 0x3e3b4a0, 0xc000118008, 0x3e70f20, 0xc0027425a0, 0x3e8bb00, 0xc0030ccc30, 0x0, 0x1, 0x0, ...)
	/home/jenkins/agent/workspace/build-tidb/tidb/planner/core/expression_rewriter.go:110 +0xb9
github.com/pingcap/tidb/planner/core.rewriteAstExpr(0x3e8c600, 0xc001fc9a00, 0x3e70f20, 0xc0027425a0, 0xc0022b6f00, 0xc0024eed00, 0x1, 0x2, 0x10, 0x36fd940, ...)
	/home/jenkins/agent/workspace/build-tidb/tidb/planner/core/expression_rewriter.go:77 +0x21d
github.com/pingcap/tidb/expression.RewriteSimpleExprWithNames(0x3e8c600, 0xc001fc9a00, 0x3e70f20, 0xc0027425a0, 0xc0022b6f00, 0xc0024eed00, 0x1, 0x2, 0x0, 0xc0017ca8f0, ...)
	/home/jenkins/agent/workspace/build-tidb/tidb/expression/simple_rewriter.go:113 +0x8e
github.com/pingcap/tidb/expression.ParseSimpleExprsWithNames(0x3e8c600, 0xc001fc9a00, 0xc002c8e732, 0xa, 0xc0022b6f00, 0xc0024eed00, 0x1, 0x2, 0x0, 0x80, ...)
	/home/jenkins/agent/workspace/build-tidb/tidb/expression/simple_rewriter.go:102 +0x352
github.com/pingcap/tidb/planner/core.generateHashPartitionExpr(0x3e8c600, 0xc001fc9a00, 0xc002f5ac80, 0xc0024ee5b0, 0x2, 0x2, 0xc0024eed00, 0x1, 0x2, 0xc0021a21e0, ...)
	/home/jenkins/agent/workspace/build-tidb/tidb/planner/core/rule_partition_processor.go:110 +0xca
github.com/pingcap/tidb/planner/core.(*partitionProcessor).findUsedPartitions(0x5cbdfe0, 0x3e8c600, 0xc001fc9a00, 0x3e87ae0, 0xc001e7fe00, 0x5cbe6a8, 0x0, 0x0, 0x0, 0x0, ...)
	/home/jenkins/agent/workspace/build-tidb/tidb/planner/core/rule_partition_processor.go:121 +0xd1
github.com/pingcap/tidb/planner/core.(*partitionProcessor).pruneHashPartition(0x5cbdfe0, 0x3e8c600, 0xc001fc9a00, 0x3e87ae0, 0xc001e7fe00, 0x5cbe6a8, 0x0, 0x0, 0x0, 0x0, ...)
	/home/jenkins/agent/workspace/build-tidb/tidb/planner/core/rule_partition_processor.go:207 +0x113
github.com/pingcap/tidb/planner/core.(*partitionProcessor).processHashPartition(0x5cbdfe0, 0xc0002c3900, 0xc002f5ac80, 0x203000, 0xc0030cc4e0, 0x1, 0x1)
	/home/jenkins/agent/workspace/build-tidb/tidb/planner/core/rule_partition_processor.go:215 +0x12f
github.com/pingcap/tidb/planner/core.(*partitionProcessor).prune(0x5cbdfe0, 0xc0002c3900, 0x203000, 0x203000, 0xc0004b8000, 0x7f2426c37b28)
	/home/jenkins/agent/workspace/build-tidb/tidb/planner/core/rule_partition_processor.go:502 +0x18e
github.com/pingcap/tidb/planner/core.(*partitionProcessor).rewriteDataSource(0x5cbdfe0, 0x3e8ac00, 0xc0002c3900, 0x1, 0x0, 0x12f61f6, 0xc0022b6d70)
	/home/jenkins/agent/workspace/build-tidb/tidb/planner/core/rule_partition_processor.go:64 +0x71d
github.com/pingcap/tidb/planner/core.(*partitionProcessor).rewriteDataSource(0x5cbdfe0, 0x3e8b600, 0xc001b9bce0, 0x1, 0xc0039c7390, 0x12f61f6, 0xc0032a7e60)
	/home/jenkins/agent/workspace/build-tidb/tidb/planner/core/rule_partition_processor.go:92 +0x62a
github.com/pingcap/tidb/planner/core.(*partitionProcessor).rewriteDataSource(0x5cbdfe0, 0x3e8b100, 0xc001b9bdc0, 0x1, 0x0, 0xc001b9bce0, 0xc0002c3900)
	/home/jenkins/agent/workspace/build-tidb/tidb/planner/core/rule_partition_processor.go:9
"]
```

### 4. What is your TiDB version? (Required)
```
Release Version: v5.0.0-rc-5-gad597a7
Edition: Community
Git Commit Hash: ad597a7ece4320988eac870b24e01d1b94c3bbaf
Git Branch: HEAD
UTC Build Time: 2021-01-28 02:24:59
GoVersion: go1.13.7
Race Enabled: false
Check Table Before Drop: false
TiKV Min Version: v3.0.0-60965b006877ca7234adaced7890d7b029ed1306
```

--- a/planner/core/partition_pruner_test.go
+++ b/planner/core/partition_pruner_test.go
@@ -461,6 +461,18 @@ CREATE TABLE t1 (
 PARTITION BY HASH( a )
 PARTITIONS 4`)
 	tk.MustQuery("SELECT (SELECT tt.a FROM t1  tt LIMIT 1) aa, COUNT(DISTINCT b) FROM t1  GROUP BY aa").Check(testkit.Rows()) // work fine without any error
+
+	tk.MustExec("insert into t1 values (1, 1)")
+	tk.MustQuery("SELECT (SELECT tt.a FROM t1  tt LIMIT 1) aa, COUNT(DISTINCT b) FROM t1  GROUP BY aa").Check(testkit.Rows("1 1"))
+
+	tk.MustExec("insert into t1 values (2, 2), (2, 2)")
+	tk.MustQuery("SELECT (SELECT tt.a FROM t1  tt LIMIT 1) aa, COUNT(DISTINCT b) FROM t1  GROUP BY aa").Check(testkit.Rows("1 2"))
+
+	tk.MustExec("insert into t1 values (3, 3), (3, 3), (3, 3)")
+	tk.MustQuery("SELECT (SELECT tt.a FROM t1  tt LIMIT 1) aa, COUNT(DISTINCT b) FROM t1  GROUP BY aa").Check(testkit.Rows("1 3"))
+
+	tk.MustExec("insert into t1 values (4, 4), (4, 4), (4, 4), (4, 4)")
+	tk.MustQuery("SELECT (SELECT tt.a FROM t1  tt LIMIT 1) aa, COUNT(DISTINCT b) FROM t1  GROUP BY aa").Check(testkit.Rows("4 4"))
 }
 
 func (s *testPartitionPruneSuit) TestIssue22898(c *C) {

--- a/planner/core/partition_pruner_test.go
+++ b/planner/core/partition_pruner_test.go
@@ -449,6 +449,20 @@ func (s *testPartitionPruneSuit) TestListColumnsPartitionPrunerRandom(c *C) {
 	}
 }
 
+func (s *testPartitionPruneSuit) TestIssue22635(c *C) {
+	tk := testkit.NewTestKit(c, s.store)
+	tk.MustExec("USE test;")
+	tk.MustExec("DROP TABLE IF EXISTS t1")
+	tk.MustExec(`
+CREATE TABLE t1 (
+  a int(11) DEFAULT NULL,
+  b int(11) DEFAULT NULL
+) ENGINE=InnoDB DEFAULT CHARSET=utf8mb4 COLLATE=utf8mb4_bin
+PARTITION BY HASH( a )
+PARTITIONS 4`)
+	tk.MustQuery("SELECT (SELECT tt.a FROM t1  tt LIMIT 1) aa, COUNT(DISTINCT b) FROM t1  GROUP BY aa").Check(testkit.Rows()) // work fine without any error
+}
+
 func (s *testPartitionPruneSuit) TestIssue22898(c *C) {
 	tk := testkit.NewTestKit(c, s.store)
 	tk.MustExec("USE test;")

--- a/planner/core/rule_partition_processor.go
+++ b/planner/core/rule_partition_processor.go
@@ -107,6 +107,7 @@ type partitionTable interface {
 
 func generateHashPartitionExpr(ctx sessionctx.Context, pi *model.PartitionInfo, columns []*expression.Column, names types.NameSlice) (expression.Expression, error) {
 	schema := expression.NewSchema(columns...)
+	fmt.Printf(">>>>>>>>>>>>>>>>>>>>>>>>> ParseSimpleExprsWithNames, pi.Expr: %v, schema: %v, names: %v \n", pi.Expr, schema.String(), names)
 	exprs, err := expression.ParseSimpleExprsWithNames(ctx, pi.Expr, schema, names)
 	if err != nil {
 		return nil, err
@@ -211,8 +212,38 @@ func (s *partitionProcessor) pruneHashPartition(ctx sessionctx.Context, tbl tabl
 	return used, nil
 }
 
+func (s *partitionProcessor) recoverTableColNames(ds *DataSource) ([]*types.FieldName, error) {
+	names := make([]*types.FieldName, 0, len(ds.TblCols))
+	colsInfo := ds.table.FullHiddenColsAndVisibleCols()
+	for _, colExpr := range ds.TblCols {
+		found := false
+		for _, colInfo := range colsInfo {
+			if colInfo.ID == colExpr.ID {
+				names = append(names, &types.FieldName{
+					DBName:      ds.DBName,
+					TblName:     ds.tableInfo.Name,
+					ColName:     colInfo.Name,
+					OrigTblName: ds.tableInfo.Name,
+					OrigColName: colInfo.Name,
+					Hidden:      colInfo.Hidden,
+				})
+				found = true
+				break
+			}
+		}
+		if !found {
+			return nil, errors.New(fmt.Sprintf("information of column %v is not found", colExpr.String()))
+		}
+	}
+	return names, nil
+}
+
 func (s *partitionProcessor) processHashPartition(ds *DataSource, pi *model.PartitionInfo) (LogicalPlan, error) {
-	used, err := s.pruneHashPartition(ds.SCtx(), ds.table, ds.partitionNames, ds.allConds, ds.TblCols, ds.names)
+	names, err := s.recoverTableColNames(ds)
+	if err != nil {
+		return nil, err
+	}
+	used, err := s.pruneHashPartition(ds.SCtx(), ds.table, ds.partitionNames, ds.allConds, ds.TblCols, names)
 	if err != nil {
 		return nil, err
 	}

--- a/planner/core/rule_partition_processor.go
+++ b/planner/core/rule_partition_processor.go
@@ -217,6 +217,10 @@ func (s *partitionProcessor) pruneHashPartition(ctx sessionctx.Context, tbl tabl
 func (s *partitionProcessor) reconstructTableColNames(ds *DataSource) ([]*types.FieldName, error) {
 	names := make([]*types.FieldName, 0, len(ds.TblCols))
 	colsInfo := ds.table.FullHiddenColsAndVisibleCols()
+	colsInfoMap := make(map[int64]*table.Column, len(colsInfo))
+	for _, c := range colsInfo {
+		colsInfoMap[c.ID] = c
+	}
 	for _, colExpr := range ds.TblCols {
 		if colExpr.ID == model.ExtraHandleID {
 			names = append(names, &types.FieldName{
@@ -227,25 +231,18 @@ func (s *partitionProcessor) reconstructTableColNames(ds *DataSource) ([]*types.
 			})
 			continue
 		}
-
-		found := false
-		for _, colInfo := range colsInfo {
-			if colInfo.ID == colExpr.ID {
-				names = append(names, &types.FieldName{
-					DBName:      ds.DBName,
-					TblName:     ds.tableInfo.Name,
-					ColName:     colInfo.Name,
-					OrigTblName: ds.tableInfo.Name,
-					OrigColName: colInfo.Name,
-					Hidden:      colInfo.Hidden,
-				})
-				found = true
-				break
-			}
+		if colInfo, found := colsInfoMap[colExpr.ID]; found {
+			names = append(names, &types.FieldName{
+				DBName:      ds.DBName,
+				TblName:     ds.tableInfo.Name,
+				ColName:     colInfo.Name,
+				OrigTblName: ds.tableInfo.Name,
+				OrigColName: colInfo.Name,
+				Hidden:      colInfo.Hidden,
+			})
+			continue
 		}
-		if !found {
-			return nil, errors.New(fmt.Sprintf("information of column %v is not found", colExpr.String()))
-		}
+		return nil, errors.New(fmt.Sprintf("information of column %v is not found", colExpr.String()))
 	}
 	return names, nil
 }

--- a/planner/core/rule_partition_processor.go
+++ b/planner/core/rule_partition_processor.go
@@ -107,7 +107,6 @@ type partitionTable interface {
 
 func generateHashPartitionExpr(ctx sessionctx.Context, pi *model.PartitionInfo, columns []*expression.Column, names types.NameSlice) (expression.Expression, error) {
 	schema := expression.NewSchema(columns...)
-	fmt.Printf(">>>>>>>>>>>>>>>>>>>>>>>>> ParseSimpleExprsWithNames, pi.Expr: %v, schema: %v, names: %v \n", pi.Expr, schema.String(), names)
 	exprs, err := expression.ParseSimpleExprsWithNames(ctx, pi.Expr, schema, names)
 	if err != nil {
 		return nil, err
@@ -216,6 +215,16 @@ func (s *partitionProcessor) recoverTableColNames(ds *DataSource) ([]*types.Fiel
 	names := make([]*types.FieldName, 0, len(ds.TblCols))
 	colsInfo := ds.table.FullHiddenColsAndVisibleCols()
 	for _, colExpr := range ds.TblCols {
+		if colExpr.ID == model.ExtraHandleID {
+			names = append(names, &types.FieldName{
+				DBName:      ds.DBName,
+				TblName:     ds.tableInfo.Name,
+				ColName:     model.ExtraHandleName,
+				OrigColName: model.ExtraHandleName,
+			})
+			continue
+		}
+
 		found := false
 		for _, colInfo := range colsInfo {
 			if colInfo.ID == colExpr.ID {

--- a/planner/core/rule_partition_processor.go
+++ b/planner/core/rule_partition_processor.go
@@ -211,7 +211,10 @@ func (s *partitionProcessor) pruneHashPartition(ctx sessionctx.Context, tbl tabl
 	return used, nil
 }
 
-func (s *partitionProcessor) recoverTableColNames(ds *DataSource) ([]*types.FieldName, error) {
+// reconstructTableColNames reconstructs FieldsNames according to ds.TblCols.
+// ds.names may not match ds.TblCols since ds.names is pruned while ds.TblCols contains all original columns.
+// please see https://github.com/pingcap/tidb/issues/22635 for more details.
+func (s *partitionProcessor) reconstructTableColNames(ds *DataSource) ([]*types.FieldName, error) {
 	names := make([]*types.FieldName, 0, len(ds.TblCols))
 	colsInfo := ds.table.FullHiddenColsAndVisibleCols()
 	for _, colExpr := range ds.TblCols {
@@ -248,7 +251,7 @@ func (s *partitionProcessor) recoverTableColNames(ds *DataSource) ([]*types.Fiel
 }
 
 func (s *partitionProcessor) processHashPartition(ds *DataSource, pi *model.PartitionInfo) (LogicalPlan, error) {
-	names, err := s.recoverTableColNames(ds)
+	names, err := s.reconstructTableColNames(ds)
 	if err != nil {
 		return nil, err
 	}

